docs: add recipe with angular elements (#1819)
ngrx/platform: projects/ngrx.io/content/guide/store/recipes/custom-elements.md

# Using Store with Angular Elements

## Creating Angular Elements

> Using Store with Angular Elements is similar to using it with standard Angular Component.

Reviewer: I would remove this statement and move up the first few sentences in the tutorial section to here. I would also use the word recipe instead of tutorial.

## Tutorial

The following tutorial shows you how to manage the state of a counter, and how to select and display it within an Angular Elmements and use the same within an Angular Project and an static page. Try the <live-example name="store-elements" downloadOnly>Anguler Elements Project</live-example>.

1. Create a new application by running the below command
```sh
ng new ngRxElementDemo --createApplication=false
```
2. Create a new Angular Library using below command
```sh
ng g library my-counter
```

> 3. Add code mentioned in [Getting Started Guide](guide/store#installation) to new created.

Reviewer: Can you go into more detail here about what is required? Also appears to be an incomplete sentence.

Reviewer: Sentence is now complete, but can you go into more depth about why they are going to the `Getting Started Guide` and what exactly they are doing?

Something like..

Follow the instructions in the `Getting Started Guide` to create a new counter store complete with `actions`, `reducer`, and `module`.

> 4. Rename `app.module.ts` to `counter.module.ts` file and add the below code.

Reviewer: Instead of just telling the reader to add code, can you go into a little more detail about what they are adding and why it is relevant?

<code-example header="src/lib/counter.module.ts" path="store-elements/projects/my-counter/src/lib/app/counter.module.ts">
</code-example>

> 5. Add 2 application within the project using the below command

Reviewer: Consider changing to “Generate two (2) new applications”

```sh
ng g application elementApp
ng g application counterelement
```

> 6. Within elementApp add the below code in `app.module.ts`.

Reviewer: Same as previously mentioned, provide some more context as to why the user is to add the code.

<code-example header="elementApp/src/app.module.ts" path="store-elements/projects/elementApp/src/app.module.ts">
</code-example>
7. Add the below code in `app.component.html`.
<code-example header="elementApp/src/app.component.html" path="store-elements/projects/elementApp/src/app.component.html">
</code-example>

> 8. Run the below command and import the code given in `polyfills.ts` file available in both the app created.

Reviewer: Consider splitting this into two steps and explaining why you need to install the web components dependency. Then explain the polyfills file and how it’s used.

Reviewer: @santoshyadav198613 this appears to be unresolved still.

```sh
npm i @webcomponents/webcomponentsjs -save
import '@webcomponents/webcomponentsjs/custom-elements-es5-adapter.js';

> ```

Reviewer: Numbers are off here. 8 is listed twice

> 8. Verify the app by runnig it locally using below command.

Reviewer: Typo “runnig” also, consider changing Verify to “Build and Serve”

```sh
ng build my-counter && ng serve — project=elementApp -o
```
9. For building an custom-element.js file which can be used outside an Angular App. Go to `counterelement` App and add the below code in `app.module.ts`.
<code-example header="counterelement/src/app.module.ts" path="store-elements/projects/counterelement/src/app.module.ts">
</code-example>
10. Run the below commands to install a package to merge files to create `custom-element.js` file.
```sh
npm install jscat --save-dev
```

> 11. Add a new folder `htmlapp` in root, create a new file in `index.html` and add the below code.

Reviewer: Create a new file “named index.html”...

<code-example header="htmlapp/index.html" path="store-elements/htmlapp/index.html">
</code-example>

> 12. Add the below command to `package.json`.

Reviewer: Change to json or js syntax highlighting and also consider wording change to “Add two (2) new commands to the “scripts” section of “package.json”

> ```sh

Reviewer: May need to change this to js or json instead of sh for code highlighting.

"build-element": "ng build my-counter && ng build --project=counterelement --prod --output-hashing=none",
"package": "jscat ./dist/counterelement/runtime.js ./dist/counterelement/polyfills.js ./dist/counterelement/scripts.js ./dist/counterelement/main.js > htmlapp/counter-element.js",
```

> 13. Run the below command to create and run the html app.

Reviewer: Consider wording like...Run the “following commands to build and package your custom elements into a single “js” file output.”

```sh
npm run build-element
npm run package
```
14. Move to `htmlapp` folder and run `http-server` make sure you have `http-server` installed globally.

> You can refer the blog post on [Angular In Depth](https://blog.angularindepth.com/using-angular-elements-with-ngrx-bc655e1eb212)

Reviewer: @timdeschryver do we want to link to blog post here?

ng g application counterelement
```
7. Run the below command to add Angular Element as a dependency to our project.
```sh
ng add @angular/elements
```
8. Once the command is completed, you will notice below changes in `angular.json`, the above package only makes changes to default project which in our case is `elementApp`, so copy the below line and add the same to the scripts section of `counterelement`
```json
"scripts": [
{"input": "node_modules/document-register-element/build/document-register-element.js"}
]
```
9. Within elementApp add the below code in `app.module.ts`.
<code-example header="elementApp/src/app/app.module.ts" path="store-elements/projects/elementApp/src/app/app.module.ts">
</code-example>
In the above code, we have imported `CounterModule` from our `my-counter` library and in `schemas` array we have added `CUSTOM_ELEMENTS_SCHEMA` because we are loading Custom Elements in our App.
10. Add the below code in `app.component.html`.
<code-example header="elementApp/src/app/app.component.html" path="store-elements/projects/elementApp/src/app/app.component.html">
</code-example>
Here we are using the Angular Elements, which we registered in our `my-counter` library.

> 11. Run the below command and import the code given in `polyfills.ts` file available in both the app created.

Reviewer: @santoshyadav198613 doesn't look like this was split into two separate commands. There is also a spacing issue as `4. Verify the app by running it locally using below command` is now on the same line as `The above polyfill...`

> 11. Run the below command and import the code given in `polyfills.ts` file available in both the app created.

Reviewer: @santoshyadav198613 `4. Verify` is still on the same line as the `The above polyfill`

> 11. Run the below command and import the code given in `polyfills.ts` file available in both the app created.

Author: Yes trying to ix, somehow it is broken as added json comment above it.

# Using Store with Angular Elements
## Creating Angular Elements

> The following recipe shows you how to manage the state of a counter, and how to select and display it within an Angular Elements and use the same within an Angular Project and a static page.

Reviewer: ```suggestion
The following recipe illustrates utilizing Angular Elements with a NgRx Store to manage the state of a counter, and select and display that counter state from within separate Angular Elements. All from within a single Angular Project and a static HTML page.
```

Reviewer: @santoshyadav198613 one more change here...

`All from within a single Angular Project and a static HTML page.`
should be
`This is all accomplished from within a single Angular Project and static HTML page.`

Author: Done.

## Recipe

> You can download  <live-example name="store-elements" downloadOnly>Angular Elements Project</live-example>.

Reviewer: Is this an alternative to following the below instructions? I'm not sure this adds any value. Might want to consider removing this as it's not clear what the reader should be doing.

Author: This is the entire example code, which can be downloaded.

Reviewer: If this is the entire code sample, then I would recommend creating a new section at the very end for readers to download the example code.

Author: makes sense.

Author: Done.

Reviewer: OK Let's take it a little further. 

Instead of just adding `You can download Angular Elements Project.` Let's create a new section with a heading called `Project Example Repository` and add a sentence about cloning the full example from GitHub with a link to the [GitHub repo](https://github.com/santoshyadav198613/platform/tree/docs--add-receipe-to-use-store-with-elelments/projects/ngrx.io/content/examples/store-elements). 

This adds more value to the reader as they can peruse the full solution on GitHub and clone for themselves to review.

Author: Ok. let me do that, it should be at the end right.

Author: But won't that be an overhead for users as they have to download the entire repo? 
By download link, we are downloading the same app, and they will get the only code they want.

1. Create a new application by running the below command
```sh
ng new ngRxElementDemo - createApplication=false
```

### Creating Library

1. Create a new Angular Library using below command
```sh
ng g library my-counter
```
2. Follow the instructions in the [Getting Started Guide](guide/store#installation) to create a new counter store complete with actions, reducer, and module.

> 3. The only difference here is split the code for increment, decrement and reset in the sepearate component.

Reviewer: ```suggestion
3. The only difference here is to split the code for increment, decrement and reset into separate components.
```

Reviewer: @santoshyadav198613 Still some typos, `reset into sepearate component.` should be `reset into separate components.`

Author: Done.

Reviewer: @santoshyadav198613 Still a typo here: `sepearate components.` should be `separate components.`

Author: Done.

4. Rename `app.module.ts` to `counter.module.ts` file and add the below code.
<code-example header="src/lib/counter.module.ts" path="store-elements/projects/my-counter/src/lib/counter.module.ts">
</code-example>

> In the above code, we have added all the components in `entryComponents` array as we need to render these components as Angular Elements.

Reviewer: ```suggestion
In the above code, we have added all the components in `entryComponents` array. This ensures the components are available for rendering as Angular Elements.
```

> Also inside `constructor` you may notice `createCustomElement`, this is used to register the Angular Components as Angular Elements.

Reviewer: ```suggestion
In addition, inside the `constructor`, `createCustomElement` must be called for each component that will be used as an Angular Element.
```

"scripts": [
{"input": "node_modules/document-register-element/build/document-register-element.js"}
]
```

### Using With Angular Application

1. Within elementApp add the below code in `app.module.ts`.
<code-example header="elementApp/src/app/app.module.ts" path="store-elements/projects/elementApp/src/app/app.module.ts">
</code-example>
In the above code, we have imported `CounterModule` from our `my-counter` library and in `schemas` array we have added `CUSTOM_ELEMENTS_SCHEMA` because we are loading Custom Elements in our App.
2. Add the below code in `app.component.html`.
<code-example header="elementApp/src/app/app.component.html" path="store-elements/projects/elementApp/src/app/app.component.html">
</code-example>
Here we are using the Angular Elements, which we registered in our `my-counter` library.
3. Run the below command and import the code given in `polyfills.ts` file available in both the app created.
```sh
npm i @webcomponents/webcomponentsjs -save
import '@webcomponents/webcomponentsjs/custom-elements-es5-adapter.js';
```
The above polyfill is required to support Custom Elements for older browser.
4. Verify the app by running it locally using below command.
```sh

> ng build my-counter && ng serve - project=elementApp -o

Reviewer: ```suggestion
ng build my-counter && ng serve --project=elementApp -o
```

```

### Using with Static Page

> 1. For building a custom-element.js file which can be used outside an Angular App. Go to `counterelement` App and add the below code in `app.module.ts`.

Reviewer: ```suggestion
1. In order to build the `custom-element.js` file, which can be used outside an Angular App, go to the `counterelement` app and add the below code in `app.module.ts`.
```

<code-example header="counterelement/src/app/app.module.ts" path="store-elements/projects/counterelement/src/app/app.module.ts">
</code-example>

> In the above code, you may notice we are not using any component inside the `bootstrap` array, as we are going to bootstrap our app manually using `ngDoBootstrap`. By making this change we will create a `counter-element.js` file which we can use inside any static page.

Reviewer: ```suggestion
In the above code, you may notice we are not using any component inside the `bootstrap` array. This is because we are going to bootstrap our app manually using `ngDoBootstrap`. This allows us to create a `counter-element.js` file which we can use inside any static page.
```

2. Run the below commands to install a package to merge files to create `custom-element.js` file.
```sh
npm install jscat - save-dev
```
3. Add a new folder `htmlapp` in the root, create a new file named `index.html` and add the below code.
<code-example header="htmlapp/index.html" path="store-elements/htmlapp/index.html">
</code-example>
4. Add two (2) new commands to the `scripts` section of `package.json`.
```json
"build-element": "ng build my-counter && ng build - project=counterelement - prod - output-hashing=none",
"package": "jscat ./dist/counterelement/runtime.js ./dist/counterelement/polyfills.js ./dist/counterelement/scripts.js ./dist/counterelement/main.js > htmlapp/counter-element.js",
```

> 5. Run the following command to build and package your custom element into a single js file.

Reviewer: ```suggestion
5. Run the following commands to build and package your custom element into a single `.js` file.
```

```sh
npm run build-element
npm run package
```

> 6. Move to `htmlapp` folder and run `http-server` make sure you have `http-server` installed globally.

Reviewer: Let's split these into separate steps for the reader. We should also provide a command snippet on how to install `http-server` globally and link to the github page for `http-server`.
